### PR TITLE
Add checkpoint code

### PR DESCRIPTION
**Pull request by cvache**
Adds code to connect to DB and store a request made to the API

Very basic and we will want to do a refactor at some point to take this from a POC to a production level function

### CODE REVIEW

File: .docker/docker-compose.yaml [hunk 1]
```diff
@@ -1,27 +1,18 @@
...
-      - backend
-
-  backend:
+  api:
     build:
-      context: flask
+      context: .
       target: dev-envs
-    stop_signal: SIGINT
+    container_name: fastapi-application
```
Comment (ethanchatfield) [inline .docker/docker-compose.yaml:6]: nit: i would lean towards fastapi-app

File: .docker/docker-compose.yaml [hunk 2]
```diff
@@ -1,27 +1,18 @@
...
-    ports:
-      - 80:80
-    depends_on:
-      - backend
-
-  backend:
+  api:
     build:
-      context: flask
+      context: .
```
Comment (ethanchatfield) [inline .docker/docker-compose.yaml:4]: what does the period mean here? ingest all context?

File: app/main.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    return {"message": "OK"}
+
+@app.post("/checkpoint")
+async def add_checkpint(checkpoint: CheckpointPostRequest):
+    #Setup database connection
+    try:
+        client = motor.motor_tornado.MotorClient("mongodb://mongo:27017/")
+        db = client.BlitzOps
+    except Exception as e:
+        return("Could not connect to database")
```
Comment (ethanchatfield) [inline app/main.py:25]: i would make this a separate function, like `init_db`

Comment (ethanchatfield) [reply app/main.py:25]: in fact, this shouldn't be in `add_checkpoint`, we shouldn't be initializing the db connection every time

File: app/main.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
+from fastapi import FastAPI
+from pydantic import BaseModel
+import motor.motor_tornado
+
+class CheckpointPostRequest(BaseModel):
+    workflow_id: str
+    run_id: str
+    checkpoint_id: str
+    timestamp: float
```
Comment (ethanchatfield) [inline app/main.py:9]: might want to consider making this a `datetime` object or something similar

File: app/main.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
+from fastapi import FastAPI
+from pydantic import BaseModel
+import motor.motor_tornado
+
+class CheckpointPostRequest(BaseModel):
+    workflow_id: str
+    run_id: str
+    checkpoint_id: str
```
Comment (ethanchatfield) [inline app/main.py:8]: changed my mind on this, i think it's fine as `checkpoint_id`

File: app/main.py [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+@app.get("/")
+def hello_world():
+    return {"message": "OK"}
+
+@app.post("/checkpoint")
+async def add_checkpint(checkpoint: CheckpointPostRequest):
+    #Setup database connection
+    try:
+        client = motor.motor_tornado.MotorClient("mongodb://mongo:27017/")
```
Comment (ethanchatfield) [inline app/main.py:22]: i think the mongo URI should be an environment variable or constant

File: .docker/docker-compose.yaml [hunk 7]
```diff
@@ -1,27 +1,18 @@
...
+      - '8000:8000'
     volumes:
       - /var/run/docker.sock:/var/run/docker.sock
     depends_on:
-      -  mongo  
+      -  mongo
+    restart: "no"
 
   mongo:
     image: mongo
```
Comment (ethanchatfield) [inline .docker/docker-compose.yaml:18]: nit: add newline

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+WORKDIR /code
+
+COPY ./requirements.txt /code/requirements.txt
+
+RUN pip install --no-cache-dir --upgrade -r /code/requirements.txt
+
+COPY ./app /code/app
+
+CMD ["uvicorn", "app.main:app", "--host", "0.0.0.0", "--port", "80"]
```
Comment (ethanchatfield) [inline Dockerfile:11]: nit: add newline

File: .gitignore [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+# Cython debug symbols
+cython_debug/
+
+# PyCharm
+#  JetBrains specific template is maintained in a separate JetBrains.gitignore that can
+#  be found at https://github.com/github/gitignore/blob/main/Global/JetBrains.gitignore
+#  and can be added to the global gitignore or merged into this file.  For a more nuclear
+#  option (not recommended) you can uncomment the following to ignore the entire idea folder.
+#.idea/
+.DS_Store
```
Comment (ethanchatfield) [inline .gitignore:161]: nit: add newline

File: app/main.py [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+    # Insert into database
+    try:
+        print(checkpoint.model_dump())
+        db_resp = await db.checkpoints.insert_one(checkpoint.model_dump())
+    except Exception as e:
+        print("Error inserting into database")
+        return(str(e))
+    
+    return str(db_resp.inserted_id)
```
Comment (ethanchatfield) [inline app/main.py:35]: nit: add newline

File: app/tester.py [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    try:
+        db_resp = checkpoints.insert_one(cpr.model_dump())
+    except Exception as e:
+        return(str(e))
+    
+    return db_resp.inserted_id
+ 
+
+if __name__ == "__main__":
+    db()
```
Comment (ethanchatfield) [inline app/tester.py:58]: nit: add newline

File: compose.yaml [hunk 12]
```diff
@@ -1,29 +1,28 @@
...
+    restart: "no"
     depends_on:
       -  mongo  
 
   mongo:
-    image: mongo
\ No newline at end of file
+    image: mongo
+    ports: 
+      - 27017:27017
```
Comment (ethanchatfield) [inline compose.yaml:28]: nit: add newline

File: app/tester.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+    except ValidationError as e:
+        return("Could not validate payload. Ensure it is formatted correctly")
+    
+    #Setup database connection
+    try:
+        client = MongoClient("mongodb://localhost:27017/")
+        db = client.BlitzOps
+        checkpoints = db.checkpoints
+    except Exception as e:
+        return("Could not connect to database")
```
Comment (ethanchatfield) [inline app/tester.py:46]: not a big deal now, but we should mock the db so we're not using the same db instance for tests

File: compose.yaml [hunk 14]
```diff
@@ -1,29 +1,28 @@
...
+      context: .
       target: builder
-    # flask requires SIGINT to stop gracefully
-    # (default stop signal from Compose is SIGTERM)
-    stop_signal: SIGINT
-    environment:
-      - FLASK_SERVER_PORT=9091
-    volumes:
-      - ./flask:/src
+    container_name: fastapi-application
```
Comment (ethanchatfield) [inline compose.yaml:18]: nit: similar to above, i'd call this fastapi-app

File: app/main.py (hunk 4 above)
Comment (cvache) [reply app/main.py:9]: It's a datetime obj on the python side but as far as the app is concerned it's just a float. Plus payload has to be json/primitives

File: app/main.py (hunk 6 above)
Comment (cvache) [reply app/main.py:22]: Yeah you right, it should be set with env

File: app/main.py (hunk 3 above)
Comment (cvache) [reply app/main.py:25]: 100% agree. Just in here because I'm too lazy to write a db init script when its only used in one function right now

**Comment by ethanchatfield** [comment]
LGTM 👍

**Review by ethanchatfield — APPROVED** [review]
LGTM 👍 once changes added

**Pull request merged**